R/ms-sql.R

```diff
@@ -0,0 +1,227 @@
+          private$interface,
+          sprintf(
+            "SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = '%s';",
+            cache_table
+          )
+        )
+        names(res) <- tolower(names(res))
+        attempt::stop_if_not(
+          nrow(res) == 2,
+          msg = "Your cache_table your only have two columns"
+        )
+        attempt::stop_if_not(
+          all(c("cache", "id") %in% res$column_name),
+          msg = "Your cache_db should have a `cache` and an `id` column."
+        )
+        attempt::stop_if_not(
+          all(c("varchar", "varbinary") %in% res$data_type),
+          msg = "Your cache_table data types should be `varchar` and `varbinary`."
+        )
+      } else {
+        DBI::dbCreateTable(
+          private$interface,
+          cache_table,
+          fields = c(
+            id = "varchar(max)",
+            cache = "varbinary(max)"
+          )
+        )
+      }
+
+      private$algo <- algo
+      private$compress <- compress
+    },
+    #' @description
+    #' Does the cache contains a given key?
+    #' @param key Name of the key.
+    #' @return TRUE/FALSE
+    has_key = function(key) {
+      res <- DBI::dbGetQuery(
+        private$interface,
+        sprintf(
+          "SELECT id FROM %s WHERE id = '%s';",
+          private$cache_table,
+          key
+        )
+      )
+      if (nrow(res) > 1) {
+        stop("Corrupted cache: more than one entry for ", key)
+      }
+      nrow(res) == 1
+    },
+    #' @description
+    #' Get a key from the cache
+    #' @param key Name of the key.
+    #' @return The value stored using the `key`
+    # Inspied by @jrosell https://stackoverflow.com/a/70288183/8236642
+    get = function(key) {
+      if (self$has_key(key)) {
+        tmp <- tempfile(fileext = ".RDS")
+        on.exit({
+          unlink(tmp, TRUE, TRUE)
+        })
+        out <- DBI::dbGetQuery(
+          private$interface,
+          sprintf(
+            "SELECT * FROM %s WHERE id = '%s';",
+            private$cache_table,
+            key
+          )
+        )
+
+        # Handling the case where the value has been deleted in-between
+        # (should be very, very corner case)
+        res <- tryCatch(
+          {
+            unserialized_out <- unserialize(out$cache[[1]])
+            writeBin(object = unserialized_out, con = tmp)
+            readRDS(tmp)
+          },
+          error = function(e) {
+            return(NULL)
+          }
+        )
+        if (is.null(res)) {
+          return(
+            structure(list(), class = "key_missing")
+          )
+        }
+        return(
+          res
+        )
+      } else {
+        return(
+          structure(list(), class = "key_missing")
+        )
+      }
+    },
+    #' @description
+    #' Set a key in the cache
+    #' @param key Name of the key.
+    #' @param value Value to store
+    #' @return Used for side effect
+    set = function(key, value) {
```

Comment:
`ms-sql$set` is the same as `postgres$set` so there is no need to redefine here (will be passed through inheritance).